charmcraft/commands/pack.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+import logging
+import os
+import pathlib
+import zipfile
+
+from charmcraft.cmdbase import BaseCommand, CommandError
+from .utils import load_yaml
+
+logger = logging.getLogger(__name__)
+
+# the minimum set of files in a bundle
+MANDATORY_FILES = {'bundle.yaml'}
+
+
+def build_zip(zippath, basedir, fpaths):
+    """Build the final file.
+
+    Note we convert all paths to str to support Python 3.5.
+    """
+    zipfh = zipfile.ZipFile(str(zippath), 'w', zipfile.ZIP_DEFLATED)
+    for fpath in fpaths:
+        zipfh.write(str(fpath), str(fpath.relative_to(basedir)))
+    zipfh.close()
+
+
+def get_paths_to_include(dirpath):
+    """Get all file/dir paths to include."""
+    allpaths = set()
+
+    # all mandatory files, which must exist (currently only bundles.yaml is mandatory, and
+    # it's verified before)
+    for fname in MANDATORY_FILES:
+        allpaths.add(dirpath / fname)
+
+    # the extra files, which must be relative
+    config = load_yaml(dirpath / 'charmcraft.yaml') or {}
+    prime_specs = config.get('parts', {}).get('bundle', {}).get('prime', [])
+
+    for spec in prime_specs:
+        if spec[0] == os.path.sep:
```

Comment:
I don't know if it's bad practice or not... but if the yaml spec requires posix, i'd use "/" instead of os.sep.  Any chance this will ever run native on windows?

Reply:
Yes, we'll run this in Windows at some point.

Reply:
"/foo" would not fail on Windows in this case.

If the user should always use POSIX pathing in the yaml, we should replace `os.path.sep` with `/`.

Or would the user be using Windows style pathing for windows series charms?

Reply:
Ah, I see your point now!!! So, I changed to `/`, and also added a comment explaining why.